iron-input.html

```diff
@@ -60,49 +68,52 @@
     },
 
     listeners: {
-      'input': '_onInput'
+      'input': '_onInput',
+      'keydown': '_onKeydown'
+    },
+
+    get patternRegExp() {
+      var pattern;
+      if (this.allowedPattern) {
+        pattern = new RegExp(this.allowedPattern);
+      } else if (this.pattern) {
+        pattern = new RegExp(this.pattern);
+      } else {
+        switch (this.type) {
+          case 'number':
+            pattern = /[0-9.,e-]/;
+            break;
+        }
+      }
+      return pattern;
     },
 
     ready: function() {
-      this._ensureValidValue();
       this.bindValue = this.value;
     },
 
     _bindValueChanged: function() {
-      // If this was called as a result of user input, then |_validateValue|
-      // has already been called in |_onInput|, and it doesn't need to be
-      // called again.
-      if (this.value != this.bindValue) {
-        this._setValueAndPreserveCaretPosition(this.bindValue);
-        this._ensureValidValue();
-      }
-
+      this.value = this.bindValue;
       // manually notify because we don't want to notify until after setting value
       this.fire('bind-value-changed', {value: this.bindValue});
     },
 
     _onInput: function() {
-      this._ensureValidValue();
+      this.bindValue = this.value;
     },
 
-    _ensureValidValue: function() {
-      var value;
-      /* rename with rc8 */
-      if (this.preventInvalidInput && !this.validate()) {
-        value = this._previousValidInput;
-      } else {
-        value = this._previousValidInput = this.value;
+    _onKeydown: function(event) {
+      if (!this.preventInvalidInput) {
+        return;
+      }
+      var regexp = this.patternRegExp;
+      if (!regexp) {
+        return;
+      }
+      var thisChar = String.fromCharCode(event.keyCode);
+      if (!regexp.test(thisChar)) {
+        event.preventDefault();
       }
-      // set value first to not call this method again from _bindValueChanged
-      this._setValueAndPreserveCaretPosition(value);
-      this.bindValue = value;
-    },
-
-    _setValueAndPreserveCaretPosition: function(value) {
```

Comment:
I don't think so because `prevent-invalid-input` is changed to work by consuming the event, instead of resetting `value`.